### PR TITLE
Run clang-tidy linter tool on the repo and fix all warnings

### PR DESCRIPTION
**Pull request by ahsonkhan**
Related to https://github.com/Azure/azure-sdk-for-c/issues/155, https://github.com/Azure/azure-sdk-for-c/issues/453, and follow up to https://github.com/Azure/azure-sdk-for-c/pull/492

https://clang.llvm.org/extra/clang-tidy/

This PR is broken down into commits that can be reviewed individually, where each commit fixes a particular type of warning, either in the header files (under `/inc`) or in the source files (`/src`).

The check `readability-inconsistent-declaration-parameter-name` is useful for catching parameter naming differences between the header and source files.

Ran on both Windows and Ubuntu 18.04 (to get platform coverage), excluding two rules: `llvm-header-guard` and `google-readability-todo`

For certain warnings, suppressed them with inline comments in source.

> clang-tidy -p . --header-filter='.*' src/azure/core/az_span.c --checks='*,-llvm-header-guard,-google-readability-todo' --list-checks -- -Iinc
```text
Enabled checks:
    abseil-duration-addition
    abseil-duration-comparison
    abseil-duration-conversion-cast
    abseil-duration-division
    abseil-duration-factory-float
    abseil-duration-factory-scale
    abseil-duration-subtraction
    abseil-duration-unnecessary-conversion
    abseil-faster-strsplit-delimiter
    abseil-no-internal-dependencies
    abseil-no-namespace
    abseil-redundant-strcat-calls
    abseil-str-cat-append
    abseil-string-find-startswith
    abseil-time-comparison
    abseil-time-subtraction
    abseil-upgrade-duration-conversions
    android-cloexec-accept
    android-cloexec-accept4
    android-cloexec-creat
    android-cloexec-dup
    android-cloexec-epoll-create
    android-cloexec-epoll-create1
    android-cloexec-fopen
    android-cloexec-inotify-init
    android-cloexec-inotify-init1
    android-cloexec-memfd-create
    android-cloexec-open
    android-cloexec-pipe
    android-cloexec-pipe2
    android-cloexec-socket
    android-comparison-in-temp-failure-retry
    boost-use-to-string
    bugprone-argument-comment
    bugprone-assert-side-effect
    bugprone-bad-signal-to-kill-thread
    bugprone-bool-pointer-implicit-conversion
    bugprone-branch-clone
    bugprone-copy-constructor-init
    bugprone-dangling-handle
    bugprone-dynamic-static-initializers
    bugprone-exception-escape
    bugprone-fold-init-type
    bugprone-forward-declaration-namespace
    bugprone-forwarding-reference-overload
    bugprone-inaccurate-erase
    bugprone-incorrect-roundings
    bugprone-infinite-loop
    bugprone-integer-division
    bugprone-lambda-function-name
    bugprone-macro-parentheses
    bugprone-macro-repeated-side-effects
    bugprone-misplaced-operator-in-strlen-in-alloc
    bugprone-misplaced-widening-cast
    bugprone-move-forwarding-reference
    bugprone-multiple-statement-macro
    bugprone-narrowing-conversions
    bugprone-not-null-terminated-result
    bugprone-parent-virtual-call
    bugprone-posix-return
    bugprone-signed-char-misuse
    bugprone-sizeof-container
    bugprone-sizeof-expression
    bugprone-string-constructor
    bugprone-string-integer-assignment
    bugprone-string-literal-with-embedded-nul
    bugprone-suspicious-enum-usage
    bugprone-suspicious-memset-usage
    bugprone-suspicious-missing-comma
    bugprone-suspicious-semicolon
    bugprone-suspicious-string-compare
    bugprone-swapped-arguments
    bugprone-terminating-continue
    bugprone-throw-keyword-missing
    bugprone-too-small-loop-variable
    bugprone-undefined-memory-manipulation
    bugprone-undelegated-constructor
    bugprone-unhandled-self-assignment
    bugprone-unused-raii
    bugprone-unused-return-value
    bugprone-use-after-move
    bugprone-virtual-near-miss
    cert-dcl03-c
    cert-dcl16-c
    cert-dcl21-cpp
    cert-dcl50-cpp
    cert-dcl54-cpp
    cert-dcl58-cpp
    cert-dcl59-cpp
    cert-env33-c
    cert-err09-cpp
    cert-err34-c
    cert-err52-cpp
    cert-err58-cpp
    cert-err60-cpp
    cert-err61-cpp
    cert-fio38-c
    cert-flp30-c
    cert-mem57-cpp
    cert-msc30-c
    cert-msc32-c
    cert-msc50-cpp
    cert-msc51-cpp
    cert-oop11-cpp
    cert-oop54-cpp
    cert-oop58-cpp
    cert-pos44-c
    clang-analyzer-apiModeling.StdCLibraryFunctions
    clang-analyzer-apiModeling.TrustNonnull
    clang-analyzer-apiModeling.google.GTest
    clang-analyzer-apiModeling.llvm.CastValue
    clang-analyzer-apiModeling.llvm.ReturnValue
    clang-analyzer-core.CallAndMessage
    clang-analyzer-core.DivideZero
    clang-analyzer-core.DynamicTypePropagation
    clang-analyzer-core.NonNullParamChecker
    clang-analyzer-core.NonnilStringConstants
    clang-analyzer-core.NullDereference
    clang-analyzer-core.StackAddrEscapeBase
    clang-analyzer-core.StackAddressEscape
    clang-analyzer-core.UndefinedBinaryOperatorResult
    clang-analyzer-core.VLASize
    clang-analyzer-core.builtin.BuiltinFunctions
    clang-analyzer-core.builtin.NoReturnFunctions
    clang-analyzer-core.uninitialized.ArraySubscript
    clang-analyzer-core.uninitialized.Assign
    clang-analyzer-core.uninitialized.Branch
    clang-analyzer-core.uninitialized.CapturedBlockVariable
    clang-analyzer-core.uninitialized.UndefReturn
    clang-analyzer-cplusplus.InnerPointer
    clang-analyzer-cplusplus.Move
    clang-analyzer-cplusplus.NewDelete
    clang-analyzer-cplusplus.NewDeleteLeaks
    clang-analyzer-cplusplus.PureVirtualCall
    clang-analyzer-cplusplus.SelfAssignment
    clang-analyzer-cplusplus.SmartPtr
    clang-analyzer-cplusplus.VirtualCallModeling
    clang-analyzer-deadcode.DeadStores
    clang-analyzer-fuchsia.HandleChecker
    clang-analyzer-nullability.NullPassedToNonnull
    clang-analyzer-nullability.NullReturnedFromNonnull
    clang-analyzer-nullability.NullabilityBase
    clang-analyzer-nullability.NullableDereferenced
    clang-analyzer-nullability.NullablePassedToNonnull
    clang-analyzer-nullability.NullableReturnedFromNonnull
    clang-analyzer-optin.cplusplus.UninitializedObject
    clang-analyzer-optin.cplusplus.VirtualCall
    clang-analyzer-optin.mpi.MPI-Checker
    clang-analyzer-optin.osx.OSObjectCStyleCast
    clang-analyzer-optin.osx.cocoa.localizability.EmptyLocalizationContextChecker
    clang-analyzer-optin.osx.cocoa.localizability.NonLocalizedStringChecker
    clang-analyzer-optin.performance.GCDAntipattern
    clang-analyzer-optin.performance.Padding
    clang-analyzer-optin.portability.UnixAPI
    clang-analyzer-osx.API
    clang-analyzer-osx.MIG
    clang-analyzer-osx.NSOrCFErrorDerefChecker
    clang-analyzer-osx.NumberObjectConversion
    clang-analyzer-osx.OSObjectRetainCount
    clang-analyzer-osx.ObjCProperty
    clang-analyzer-osx.SecKeychainAPI
    clang-analyzer-osx.cocoa.AtSync
    clang-analyzer-osx.cocoa.AutoreleaseWrite
    clang-analyzer-osx.cocoa.ClassRelease
    clang-analyzer-osx.cocoa.Dealloc
    clang-analyzer-osx.cocoa.IncompatibleMethodTypes
    clang-analyzer-osx.cocoa.Loops
    clang-analyzer-osx.cocoa.MissingSuperCall
    clang-analyzer-osx.cocoa.NSAutoreleasePool
    clang-analyzer-osx.cocoa.NSError
    clang-analyzer-osx.cocoa.NilArg
    clang-analyzer-osx.cocoa.NonNilReturnValue
    clang-analyzer-osx.cocoa.ObjCGenerics
    clang-analyzer-osx.cocoa.RetainCount
    clang-analyzer-osx.cocoa.RetainCountBase
    clang-analyzer-osx.cocoa.RunLoopAutoreleaseLeak
    clang-analyzer-osx.cocoa.SelfInit
    clang-analyzer-osx.cocoa.SuperDealloc
    clang-analyzer-osx.cocoa.UnusedIvars
    clang-analyzer-osx.cocoa.VariadicMethodTypes
    clang-analyzer-osx.coreFoundation.CFError
    clang-analyzer-osx.coreFoundation.CFNumber
    clang-analyzer-osx.coreFoundation.CFRetainRelease
    clang-analyzer-osx.coreFoundation.containers.OutOfBounds
    clang-analyzer-osx.coreFoundation.containers.PointerSizedValues
    clang-analyzer-security.FloatLoopCounter
    clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling
    clang-analyzer-security.insecureAPI.SecuritySyntaxChecker
    clang-analyzer-security.insecureAPI.UncheckedReturn
    clang-analyzer-security.insecureAPI.bcmp
    clang-analyzer-security.insecureAPI.bcopy
    clang-analyzer-security.insecureAPI.bzero
    clang-analyzer-security.insecureAPI.decodeValueOfObjCType
    clang-analyzer-security.insecureAPI.getpw
    clang-analyzer-security.insecureAPI.gets
    clang-analyzer-security.insecureAPI.mkstemp
    clang-analyzer-security.insecureAPI.mktemp
    clang-analyzer-security.insecureAPI.rand
    clang-analyzer-security.insecureAPI.strcpy
    clang-analyzer-security.insecureAPI.vfork
    clang-analyzer-unix.API
    clang-analyzer-unix.DynamicMemoryModeling
    clang-analyzer-unix.Malloc
    clang-analyzer-unix.MallocSizeof
    clang-analyzer-unix.MismatchedDeallocator
    clang-analyzer-unix.Vfork
    clang-analyzer-unix.cstring.BadSizeArg
    clang-analyzer-unix.cstring.CStringModeling
    clang-analyzer-unix.cstring.NullArg
    clang-analyzer-valist.CopyToSelf
    clang-analyzer-valist.Uninitialized
    clang-analyzer-valist.Unterminated
    clang-analyzer-valist.ValistBase
    cppcoreguidelines-avoid-c-arrays
    cppcoreguidelines-avoid-goto
    cppcoreguidelines-avoid-magic-numbers
    cppcoreguidelines-c-copy-assignment-signature
    cppcoreguidelines-explicit-virtual-functions
    cppcoreguidelines-init-variables
    cppcoreguidelines-interfaces-global-init
    cppcoreguidelines-macro-usage
    cppcoreguidelines-narrowing-conversions
    cppcoreguidelines-no-malloc
    cppcoreguidelines-non-private-member-variables-in-classes
    cppcoreguidelines-owning-memory
    cppcoreguidelines-pro-bounds-array-to-pointer-decay
    cppcoreguidelines-pro-bounds-constant-array-index
    cppcoreguidelines-pro-bounds-pointer-arithmetic
    cppcoreguidelines-pro-type-const-cast
    cppcoreguidelines-pro-type-cstyle-cast
    cppcoreguidelines-pro-type-member-init
    cppcoreguidelines-pro-type-reinterpret-cast
    cppcoreguidelines-pro-type-static-cast-downcast
    cppcoreguidelines-pro-type-union-access
    cppcoreguidelines-pro-type-vararg
    cppcoreguidelines-slicing
    cppcoreguidelines-special-member-functions
    darwin-avoid-spinlock
    darwin-dispatch-once-nonstatic
    fuchsia-default-arguments-calls
    fuchsia-default-arguments-declarations
    fuchsia-header-anon-namespaces
    fuchsia-multiple-inheritance
    fuchsia-overloaded-operator
    fuchsia-restrict-system-includes
    fuchsia-statically-constructed-objects
    fuchsia-trailing-return
    fuchsia-virtual-inheritance
    google-build-explicit-make-pair
    google-build-namespaces
    google-build-using-namespace
    google-default-arguments
    google-explicit-constructor
    google-global-names-in-headers
    google-objc-avoid-nsobject-new
    google-objc-avoid-throwing-exception
    google-objc-function-naming
    google-objc-global-variable-declaration
    google-readability-avoid-underscore-in-googletest-name
    google-readability-braces-around-statements
    google-readability-casting
    google-readability-function-size
    google-readability-namespace-comments
    google-runtime-int
    google-runtime-operator
    google-runtime-references
    google-upgrade-googletest-case
    hicpp-avoid-c-arrays
    hicpp-avoid-goto
    hicpp-braces-around-statements
    hicpp-deprecated-headers
    hicpp-exception-baseclass
    hicpp-explicit-conversions
    hicpp-function-size
    hicpp-invalid-access-moved
    hicpp-member-init
    hicpp-move-const-arg
    hicpp-multiway-paths-covered
    hicpp-named-parameter
    hicpp-new-delete-operators
    hicpp-no-array-decay
    hicpp-no-assembler
    hicpp-no-malloc
    hicpp-noexcept-move
    hicpp-signed-bitwise
    hicpp-special-member-functions
    hicpp-static-assert
    hicpp-undelegated-constructor
    hicpp-uppercase-literal-suffix
    hicpp-use-auto
    hicpp-use-emplace
    hicpp-use-equals-default
    hicpp-use-equals-delete
    hicpp-use-noexcept
    hicpp-use-nullptr
    hicpp-use-override
    hicpp-vararg
    linuxkernel-must-check-errs
    llvm-include-order
    llvm-namespace-comment
    llvm-prefer-isa-or-dyn-cast-in-conditionals
    llvm-prefer-register-over-unsigned
    llvm-qualified-auto
    llvm-twine-local
    misc-definitions-in-headers
    misc-misplaced-const
    misc-new-delete-overloads
    misc-non-copyable-objects
    misc-non-private-member-variables-in-classes
    misc-redundant-expression
    misc-static-assert
    misc-throw-by-value-catch-by-reference
    misc-unconventional-assign-operator
    misc-uniqueptr-reset-release
    misc-unused-alias-decls
    misc-unused-parameters
    misc-unused-using-decls
    modernize-avoid-bind
    modernize-avoid-c-arrays
    modernize-concat-nested-namespaces
    modernize-deprecated-headers
    modernize-deprecated-ios-base-aliases
    modernize-loop-convert
    modernize-make-shared
    modernize-make-unique
    modernize-pass-by-value
    modernize-raw-string-literal
    modernize-redundant-void-arg
    modernize-replace-auto-ptr
    modernize-replace-random-shuffle
    modernize-return-braced-init-list
    modernize-shrink-to-fit
    modernize-unary-static-assert
    modernize-use-auto
    modernize-use-bool-literals
    modernize-use-default-member-init
    modernize-use-emplace
    modernize-use-equals-default
    modernize-use-equals-delete
    modernize-use-nodiscard
    modernize-use-noexcept
    modernize-use-nullptr
    modernize-use-override
    modernize-use-trailing-return-type
    modernize-use-transparent-functors
    modernize-use-uncaught-exceptions
    modernize-use-using
    mpi-buffer-deref
    mpi-type-mismatch
    objc-avoid-nserror-init
    objc-forbidden-subclassing
    objc-missing-hash
    objc-property-declaration
    objc-super-self
    openmp-exception-escape
    openmp-use-default-none
    performance-faster-string-find
    performance-for-range-copy
    performance-implicit-conversion-in-loop
    performance-inefficient-algorithm
    performance-inefficient-string-concatenation
    performance-inefficient-vector-operation
    performance-move-const-arg
    performance-move-constructor-init
    performance-no-automatic-move
    performance-noexcept-move-constructor
    performance-trivially-destructible
    performance-type-promotion-in-math-fn
    performance-unnecessary-copy-initialization
    performance-unnecessary-value-param
    portability-simd-intrinsics
    readability-avoid-const-params-in-decls
    readability-braces-around-statements
    readability-const-return-type
    readability-container-size-empty
    readability-convert-member-functions-to-static
    readability-delete-null-pointer
    readability-deleted-default
    readability-else-after-return
    readability-function-size
    readability-identifier-naming
    readability-implicit-bool-conversion
    readability-inconsistent-declaration-parameter-name
    readability-isolate-declaration
    readability-magic-numbers
    readability-make-member-function-const
    readability-misleading-indentation
    readability-misplaced-array-index
    readability-named-parameter
    readability-non-const-parameter
    readability-qualified-auto
    readability-redundant-access-specifiers
    readability-redundant-control-flow
    readability-redundant-declaration
    readability-redundant-function-ptr-dereference
    readability-redundant-member-init
    readability-redundant-preprocessor
    readability-redundant-smartptr-get
    readability-redundant-string-cstr
    readability-redundant-string-init
    readability-simplify-boolean-expr
    readability-simplify-subscript-expr
    readability-static-accessed-through-instance
    readability-static-definition-in-anonymous-namespace
    readability-string-compare
    readability-uniqueptr-delete-release
    readability-uppercase-literal-suffix
    zircon-temporary-objects
```

On Ubuntu (using clang-tidy v9), from the `<repo root>/sdk/` directory:
```bash
find inc/ -name '*.h' -exec clang-tidy-9 -p . {} --checks='*,-llvm-header-guard,-google-readability-todo' -- -Iinc \; > output_header_filtered.txt
find src/ -name '*.c' -exec clang-tidy-9 -p . {} --checks='*,-llvm-header-guard,-google-readability-todo' -- -Iinc \; > output_source_filtered.txt
find src/ -name '*.c' -exec clang-tidy-9 -p . --header-filter='.*' {} --checks='*,-llvm-header-guard,-google-readability-todo' -- -Iinc \; > output_with_header_filter_star.txt
```

On Windows (using clang-tidy v10), from the `<repo root>/sdk/` directory:
```bash
find inc/ -name '*.h' -exec clang-tidy -p . {} --checks='*,-llvm-header-guard,-google-readability-todo' -- -Iinc \; > output_header_filtered-windows.txt
find src/ -name '*.c' -exec clang-tidy -p . {} --checks='*,-llvm-header-guard,-google-readability-todo' -- -Iinc \; > output_source_filtered-windows.txt
find src/ -name '*.c' -exec clang-tidy -p . --header-filter='.*' {} --checks='*,-llvm-header-guard,-google-readability-todo' -- -Iinc \; > output_with_header_filter_star-windows.txt
```

### CODE REVIEW

File: sdk/inc/azure/core/internal/az_precondition_internal.h [hunk 1]
```diff
@@ -117,8 +117,8 @@ AZ_NODISCARD AZ_INLINE bool _az_span_is_valid(az_span span, int32_t min_size, bo
   // left until the theoretical end of the address space, it is not a valid span.
   // Example: (az_span) { .ptr = (uint8_t*)(~0 - 5), .size = 10 } is not a valid span, because most
   // likely you end up pointing to 0x0000 at .ptr[6], &.ptr[7] is 0x...0001, etc.
-  uint8_t* const max_ptr = (uint8_t*)~0;
-  result &= ((size_t)span_size <= (size_t)(max_ptr - ptr));
+  uint8_t* const max_ptr = (uint8_t*)-1;
```
Comment (vhvb1989) [inline sdk/inc/azure/core/internal/az_precondition_internal.h:120]: it looks weird to use signed numbers for unsigned variables. Can we do left shift to `1` with size of uint8_t?
what's the issue with `~`?   would it work by doing 
```
uint_t const value = ~0;
uint8_t* max_ptr = &value;
```

We don't really need the value to live out of the function right?

File: sdk/src/azure/core/az_hex_private.h [hunk 2]
```diff
@@ -19,6 +19,7 @@ enum
  */
 AZ_NODISCARD AZ_INLINE uint8_t _az_number_to_upper_hex(uint8_t number)
 {
+  // NOLINTNEXTLINE(readability-magic-numbers, cppcoreguidelines-avoid-magic-numbers)
```
Comment (vhvb1989) [inline sdk/src/azure/core/az_hex_private.h:22]: add an enum as well?  
just like for the other fixes

File: sdk/src/azure/core/az_json_writer.c [hunk 3]
```diff
@@ -829,26 +825,18 @@ static AZ_NODISCARD az_result _az_json_writer_append_literal(
     remaining_json = az_span_copy_u8(remaining_json, ',');
   }
 
-  remaining_json = az_span_copy(remaining_json, literal);
+  az_span_copy(remaining_json, literal);
```
Comment (vhvb1989) [inline sdk/src/azure/core/az_json_writer.c:828]: I think all non-void functions should be AZ_NODISCARD.
Then if value is return value is not used again and that's a warning, we should explicitly add `void(variable)` so it is clear about the ignore for the variable.
No need to change it, however, it's really not too much trouble

File: sdk/src/azure/core/az_json_writer.c [hunk 4]
```diff
@@ -829,26 +825,18 @@ static AZ_NODISCARD az_result _az_json_writer_append_literal(
...
-  else
-  {
-    result = _az_json_writer_append_literal(
-        ref_json_writer, AZ_SPAN_FROM_STR("false"), AZ_JSON_TOKEN_FALSE);
-  }
-  return result;
+  return value ? _az_json_writer_append_literal(
+             ref_json_writer, AZ_SPAN_FROM_STR("true"), AZ_JSON_TOKEN_TRUE)
+               : _az_json_writer_append_literal(
+                   ref_json_writer, AZ_SPAN_FROM_STR("false"), AZ_JSON_TOKEN_FALSE);
```
Comment (vhvb1989) [inline sdk/src/azure/core/az_json_writer.c:839]: why is this better than before?
looks easier to read and follow like before.
Is there another way to fix this?  What about assigning an initial value to `result` ?  
Or calling return inside each condition?

File: sdk/src/azure/core/az_span.c [hunk 5]
```diff
@@ -382,13 +382,15 @@ AZ_NODISCARD az_result az_span_atod(az_span source, double* out_number)
 
   // Stack based string to allow thread-safe mutation.
   // The length is 8 to allow space for the null-terminating character.
+  // NOLINTNEXTLINE(readability-magic-numbers,  cppcoreguidelines-avoid-magic-numbers)
   char format[8] = "%00lf%n";
```
Comment (vhvb1989) [inline sdk/src/azure/core/az_span.c:386]: why not adding the macro or enum for the 8 instead?

**Review by vhvb1989 — APPROVED** [review]
Added  a few comments.
Not that really important.

File: sdk/src/azure/core/az_json_reader.c [hunk 6]
```diff
@@ -271,7 +271,8 @@ AZ_NODISCARD static az_result _az_json_reader_process_string(az_json_reader* ref
     {
       break;
     }
-    else if (next_byte == '\\')
+
```
Comment (RickWinter) [inline sdk/src/azure/core/az_json_reader.c:274]: Was changing the `else if` to `if` intended?  Is there a test case for this bug?

File: sdk/inc/azure/core/az_span.h [hunk 7]
```diff
@@ -303,6 +303,7 @@ az_span az_span_copy_u8(az_span destination, uint8_t byte);
  */
 AZ_INLINE void az_span_fill(az_span destination, uint8_t value)
 {
+  // NOLINTNEXTLINE(clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling)
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/inc/azure/core/az_span.h:306]: The deprecation is not about memset vs memcpy, it is memset vs memset_s and memcpy vs memcpy_s. We can't use the `_s` variants because those are not available on C99.

Here's the full warning we are suppressing:
> warning: Call to function 'memset' is insecure as it does not provide security checks introduced in the C11 standard. Replace with analogous functions that support length arguments or provides boundary checks such as 'memset_s' in case of C11 

https://en.cppreference.com/w/c/string/byte/memset

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 1 above)
Comment (ahsonkhan) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:120]: > what's the issue with `~`?   would it work by doing 

No, it won't.

The issue is with `~0`. We are using a binary negation operation on a signed int `0` (`int` is the  default type for number literals).

I think `~0U` might work. I will try that or use bit-shift instead of negation (however that has the same issue with signing)

File: sdk/src/azure/core/az_json_reader.c (hunk 6 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_reader.c:274]: Yes, it was intentional. Yes, we have test coverage. No, it isn't a bug.

This is in no way changing the execution/behavior of the code. It is merely a readability fix since we don't need an `else if` following a `break`, similar to how we don't need `else if` after an `if` block that `returns`.

It's this check: `readability-else-after-return`

File: sdk/src/azure/core/az_span.c (hunk 5 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_span.c:386]: I intentionally avoided adding named symbols for things that are one-off, or didn't have a useful name outside of the local instance. If desired, we can add them later.

In this case, it is the length of the format string. If one changes, we have to change the other. Putting a named constant far away hurts maintainability, so not worth the trade-off. Keeping these localized next to each other is valuable.

File: sdk/src/azure/core/az_json_writer.c (hunk 4 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_writer.c:839]: Yes, there are other ways to fix this (like you suggested). Given this method is fairly simple, I like the conciseness of the ternary operation here.

The issue with multiple returns is in non-optimized compilers, there is more code size bloat.

File: sdk/src/azure/core/az_json_writer.c (hunk 3 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_writer.c:828]: > I think all non-void functions should be AZ_NODISCARD.

I am not sure I agree with doing this for _all_.

For example, in this case, it is totally reasonable for the user to not care about the remaining slice, especially after a call to a null terminating az_span_copy_u8. We shouldn't force the caller to jump through hoops when not necessary.

File: sdk/src/azure/core/az_hex_private.h (hunk 2 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_hex_private.h:22]: I intentionally avoided adding named symbols for things that are one-off, or didn't have a useful name outside of the local instance. If desired, we can add them later.

Got a suggestion for the name here? The 10 here is essentially one more than the last non-alphabetic digit of the hex values (0-9, A-F).
I couldn't come up with something useful, and this is not being shared anywhere else, so left it as is.

File: sdk/inc/azure/core/internal/az_http_internal.h [hunk 8]
```diff
@@ -11,6 +11,12 @@
 
 #include <azure/core/_az_cfg_prefix.h>
 
+enum
+{
+  /// The maximum number of HTTP pipeline policies allowed.
+  _az_MAXIMUM_NUMBER_OF_POLICIES = 10,
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/inc/azure/core/internal/az_http_internal.h:17]: We follow this pattern elsewhere already.

The guidelines suggest to use enums over macro and const:
https://azure.github.io/azure-sdk/clang_implementation.html#clang-no-magic-numbers
> Prefer enum values since the debugger can display the label and value and no memory is allocated. If you use const, memory is allocated. If you use #define, the debugger cannot display the label.

File: sdk/inc/azure/core/az_span.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (vhvb1989) [reply sdk/inc/azure/core/az_span.h:306]: yes, so, we can leave the warning disable for memmove and replace this memset for a loop routine that calls az_copy to set each destination.

Same thing at the end.. just one less warning to be ignored

File: sdk/src/azure/core/az_hex_private.h (hunk 2 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_hex_private.h:22]: maybe just add this explanation about what 10 means as a comment and that's fine

File: sdk/src/azure/core/az_json_writer.c (hunk 3 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_json_writer.c:828]: isn't the reason of asking the NODISCARD that the function wants to make sure that it is the user the one who might ignore the result but never the function itself?

I agree that it is totally reasonable for the user to not care about the remaining slice, but I believe it is the user who should take the call by getting getting the result and explicitly ignoring it.

I'm saying that all non-void functions in the SDK should follow the same pattern..  either letting user implicitly ignore the function return or always doing the NODISCARD.

But I don't really care that much at the end

File: sdk/src/azure/core/az_json_reader.c [hunk 9]
```diff
@@ -164,7 +164,7 @@ AZ_NODISCARD static az_result _az_json_reader_get_next_buffer(
 
 AZ_NODISCARD static az_span _az_json_reader_skip_whitespace(az_json_reader* ref_json_reader)
 {
-  az_span json = { 0 };
+  az_span json;
```
Comment (antkmsft) [inline sdk/src/azure/core/az_json_reader.c:167]: do we have to remove this? What I like about initializations, is that in case there's a bug, the behavior is consistent, so it shows up every time, rather than randomly.

**Review by antkmsft — APPROVED** [review]
I love the changes, some are fixing potential problems (brackets in macros), others simplify the code (ifs and return values simplifications). I am not sure about removing the `= { 0 }` initializations (if anything, I would prefer to have everything initialized, and rather explicitly initialized; optimizing compiler would throw it away if it sees the variable gets initialized in all cases), and I am surprised we have to rewrite unsigned `~0` as `-1`, when `~0` is what we really want; the fact that `-1` and `~0` do really have the same binary value is true on two's compliment systems, and I am not arguing that we should support systems with different negative numbers representation, but there is some indirection between saying that we want a uint with all binary digits set to `1`s, vs saying it the other way around, that we want `-1` cast to a uint.

File: sdk/src/azure/core/az_json_reader.c (hunk 9 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_json_reader.c:167]: I fixed all instances of uninitialized values that the linter picked up.
Unlike the usual culprit which is common, this scenario is different.

A lot of times we have this scenario (`out_` parameters, for example), and I totally agree we should initialize the local here (the linter catches this).
**Bad:**
```C
int x;
some_method(..., &x);
```
**Good:**
```C
int x  = 0;
some_method(..., &x);
```

However, the linter also picks up redundant assignment, which is the precise case here and why I removed the initialization. In this case, we shouldn't be unnecessarily assigning a value that we aren't reading (`clang-analyzer-deadcode.DeadStore`).
**Bad:**
```C
int x = 0;
x = 5;
```
**Good:**
```C
int x;
x = 5;
```

File: sdk/inc/azure/core/az_span.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/inc/azure/core/az_span.h:306]: I prefer to leave things as is. I am not sure that the perf characteristics are identical. Feel free to try it out and submit a fix in a separate PR.

File: sdk/src/azure/core/az_json_reader.c (hunk 9 above)
Comment (momuno) [reply sdk/src/azure/core/az_json_reader.c:167]: This is interesting.  I know there are some different views on the team with regards to initializing a variable at definition before passing it to a method that would initialize it.  (Non primitives like structs are different... they ought to use an init function and not have all member variables initialized individually before passing to an init function.)

@danewalton has a perspective on the advantage of gaining compiler warnings by not initializing at definition.

Another exception I've seen to this is the non-initialization of a return code or result at definition.  The advantage here is to avoid a developer error of not changing the variable to the correct value before returning it.  However, I can see this as a logic error that should be caught by a unit test.

Thoughts?

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 1 above)
Comment (momuno) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:120]: I recall seeing ~0U used before.  Did it work?

File: sdk/inc/azure/core/internal/az_precondition_internal.h [hunk 10]
```diff
@@ -117,8 +117,8 @@ AZ_NODISCARD AZ_INLINE bool _az_span_is_valid(az_span span, int32_t min_size, bo
...
+  result = result && ((size_t)span_size <= (size_t)(max_ptr - ptr));
```
Comment (momuno) [inline sdk/inc/azure/core/internal/az_precondition_internal.h:121]: Why did you change from a bitwise & to a logical &&?

Comment (ahsonkhan) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:121]: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]

Also https://stackoverflow.com/a/24560

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 1 above)
Comment (ahsonkhan) [reply sdk/inc/azure/core/internal/az_precondition_internal.h]: > I recall seeing ~0U used before in other code I have worked on outside of Microsoft. Did it work?

Yes, ~0U worked.

File: sdk/inc/azure/core/internal/az_precondition_internal.h (hunk 10 above)
Comment (momuno) [reply sdk/inc/azure/core/internal/az_precondition_internal.h:121]: ah, yes I see those are both booleans now.  definitely more deterministic now. thanks.
